pkg/network/mycelium/mycelium.go

```diff
@@ -0,0 +1,235 @@
+		}
+		if err := z.Forget(zinitService); err != nil && !errors.Is(err, zinit.ErrUnknownService) {
+			return errors.Wrap(err, "failed to forget mycelium service")
+		}
+	}
+
+	bin, err := exec.LookPath(myBin)
+	if err != nil {
+		return err
+	}
+
+	cmd := `sh -c '
+    exec ip netns exec %s %s --key-file %s --tun-name %s --peers %s
+    '`
+
+	err = zinit.AddService(zinitService, zinit.InitService{
+		Exec: fmt.Sprintf(cmd, ns, bin, confPath, s.cfg.TunName, strings.Join(s.cfg.Peers, " ")),
+		After: []string{
+			"node-ready",
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := z.Monitor(zinitService); err != nil && !errors.Is(err, zinit.ErrAlreadyMonitored) {
+		return err
+	}
+
+	return z.StartWait(time.Second*20, zinitService)
+}
+
+// NodeID returns mycelium node ID of s
+func (s *MyServer) NodeID() (ed25519.PublicKey, error) {
+	if s.cfg.PublicKey == "" {
+		panic("EncryptionPublicKey empty")
+	}
+
+	return hex.DecodeString(s.cfg.PublicKey)
+}
+
+func (s *MyServer) InspectMycelium() (inspection MyceliumInspection, err error) {
+	// we check if the file exists before we do inspect because mycelium will create a random seed
+	// file if file does not exist
+	_, err = os.Stat(confPath)
+	if err != nil {
+		return inspection, err
+	}
+
+	bin, err := exec.LookPath(myBin)
+	if err != nil {
+		return inspection, err
+	}
+
+	cmd := fmt.Sprintf(`exec ip netns exec %s %s inspect --json --key-file %s `, s.ns, bin, confPath)
+	output, err := exec.Command("sh", "-c", cmd).Output()
+	if err != nil {
+		return inspection, errors.Wrap(err, "failed to inspect mycelium ip")
+	}
+
+	if err := json.Unmarshal(output, &inspection); err != nil {
+		return inspection, errors.Wrap(err, "failed to load mycelium information from key")
+	}
+
+	return
+}
+
+// IP return the address in the 400::/7 subnet allocated by mycelium
+func (m *MyceliumInspection) IP() net.IP {
+	return net.IP(m.Address)
+}
+
+// Subnet return the 400::/64 subnet allocated by mycelium
+func (m *MyceliumInspection) Subnet() (subnet net.IPNet, err error) {
+	ipv6 := m.Address.To16()
+	if ipv6 == nil {
+		return subnet, errors.Errorf("invalid mycelium ip")
+	}
+
+	ip := make(net.IP, net.IPv6len)
+	copy(ip[0:8], ipv6[0:8])
+
+	subnet = net.IPNet{
+		IP:   ip,
+		Mask: net.CIDRMask(64, 128),
+	}
+
+	return
+}
+
+// Gateway derive the gateway IP from the mycelium IP in the /64 range.
+func (m *MyceliumInspection) Gateway() (gw net.IPNet, err error) {
+	subnet, err := m.Subnet()
+	if err != nil {
+		return gw, err
+	}
+
+	ip := subnet.IP
+	ip[net.IPv6len-1] = 1
+
+	gw = net.IPNet{
+		IP:   ip,
+		Mask: net.CIDRMask(64, 128),
+	}
+
+	return
+}
+
+// Tun return the name of the TUN interface created by mycelium
+func (s *MyServer) Tun() string {
+	return s.cfg.TunName
+}
+
+// SubnetFor return an IP address out of the node allocated subnet by hasing b and using it
+// to generate the last 64 bits of the IPV6 address
+func (m *MyceliumInspection) SubnetFor(b []byte) (net.IPNet, error) {
```

Comment:
I really think this should be called `IPFor` not `SubnetFor` even if it's called like that in ygg it's still wrong.